### PR TITLE
Add promote operands op to transform dialect

### PR DESCRIPTION
**Pull request by harsh-nod**
This op promotes the given target's operands specified by the indices.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 1]
```diff
@@ -655,5 +656,33 @@ transform_dialect::VectorToMMAConversionOp::applyToOne(
...
 }
 
+DiagnosedSilenceableFailure transform_dialect::PromoteOperandsOp::applyToOne(
+    Operation *target, transform::ApplyToEachResultList &results,
+    transform::TransformState &state) {
+  OpBuilder builder(getContext());
+  OpBuilder::InsertionGuard g(builder);
+  builder.setInsertionPoint(target);
+  SmallVector<int64_t> indices = llvm::to_vector(getIndices());
+  auto numOperands = target->getNumOperands();
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:666]: nit: please spell out the type.
(the rule we follow is here: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable)

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 2]
```diff
@@ -655,5 +656,33 @@ transform_dialect::VectorToMMAConversionOp::applyToOne(
   return DiagnosedSilenceableFailure::success();
 }
 
+DiagnosedSilenceableFailure transform_dialect::PromoteOperandsOp::applyToOne(
+    Operation *target, transform::ApplyToEachResultList &results,
+    transform::TransformState &state) {
+  OpBuilder builder(getContext());
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:662]: I believe we want to use `  IRRewriter rewriter(getContext());` for better tracking

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 3]
```diff
@@ -655,5 +656,33 @@ transform_dialect::VectorToMMAConversionOp::applyToOne(
...
+  SmallVector<int64_t> indices = llvm::to_vector(getIndices());
+  auto numOperands = target->getNumOperands();
+  bufferization::BufferizationOptions options;
+  for (int64_t index : indices) {
+    if ((index >= 0) && (index < numOperands)) {
+      FailureOr<Value> ret = bufferization::allocateTensorForShapedValue(
+          builder, target->getLoc(), target->getOperand(index), false, options,
+          true);
+      if (failed(ret)) {
+        target->emitOpError("failed to promote operand");
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:674]: I think you want:
`return emitDefaultDefiniteFailure(target) << "failed to promote operand";` same below.

**Review by ThomasRaoux — APPROVED** [review]
Thanks!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp (hunk 1 above)
Comment (harsh-nod) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:666]: sure will change

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp (hunk 2 above)
Comment (harsh-nod) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:662]: sounds good. will modify.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp (hunk 3 above)
Comment (harsh-nod) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:674]: thanks will change

**Comment by hanhanW** [comment]
There are failures, maybe it's because of mid air collision

**Comment by hanhanW** [comment]
I think https://github.com/iree-org/iree/pull/12056 fixes the issue.

**Pull request merged**